runtimebp/cpu.go

```diff
@@ -127,11 +131,83 @@ func defaultMaxProcsFormula(n float64) int {
 	return int(math.Ceil(n))
 }
 
-// GOMAXPROCS sets runtime.GOMAXPROCS with the default formula,
+func scaledMaxProcsFormula(n float64) int {
+	i := fetchMaxProcsMult()
+	return int(math.Ceil(n * float64(i)))
+}
+
+func fetchMaxProcsMult() int {
+	// Allow using a multiplier for number of processes relative to limit.
+	// Not catching the ok here because this function will never be called unless
+	// the environment variable is set
+	v, _ := os.LookupEnv("GOMAXPROCSMULT")
+	i, err := strconv.Atoi(v)
+	if err != nil {
+		fmt.Fprintf(
+			os.Stderr,
+			"runtimebp.fetchMaxProcsMult: GOMAXPROCSMULT is set to %q, which is invalid, falling back to 2 to ensure parallelism: %v",
+			v,
+			err,
+		)
+		i = 2
+	}
+	if i <= 0 {
+		i = 1
+	}
+	return i
+}
+
+// fetchCpuRequest checks for the magic CPUREQUEST variable set as a fallback
+// by Infrared.  If this is set, we use it and the multiplier to set our
+// GOMAXPROCS as a fallback.
+func fetchCPURequest() (int, error) {
+	if v, ok := os.LookupEnv("CPUREQUEST"); ok {
+		req, err := strconv.Atoi(v)
+		if err != nil {
+			// This case will probably never be hit, but I'm keeping it just in case.
+			// In k8s 1.21 and newer, when you present the CPU request via the Downward API
+			// k8s automatically rounds it up to the nearest integer unit.  That's how
+			// we plan to set this variable, but just in case it might be worth doing this
+			// check in case they ever break that.  There's no official contract for that.
+			match := millicoreRegexp.FindStringSubmatch(v)
+			if match == nil {
+				fmt.Fprintf(
+					os.Stderr,
+					"runtimebp.fetchCPURequest: CPUREQUEST is set to %q, which is invalid, ignoring.",
+					v,
+				)
+				req = 1
+			} else {
+				m, err := strconv.Atoi(match[0])
+				if err != nil {
+					req = int(math.Ceil(float64(m) / 1000.0))
```

Comment:
:no_bell:/super nit: in go you don't have to add `.0`.

Reply:
:no_bell: also I'm wondering if it would make more sense to keep `req` as a float64 when passing it into `scaledMaxProcsFormula`, vs. ceil it here.

Reply:
I can drop the `.0` no prob, but the value generally comes in as an int, so otherwise I think I'd have to do a bunch of hoop jumping in reverse.

Reply:
isn't the whole point of this part to handle non-int cpu requests?

so what I'm suggesting is:
1. define `var req float64` at the beginning of this function
2. in the `ok` branch, use a local variable to get the converted int value from `strconv.Atoi`, and cast+assign it to `req` if the conversion was successful
3. here we just do `req = float64(m) / 1000` without ceil nor converting to int.

Reply:
Right, but it still would be coming in as an int in the string.  Like, if it came in as `500m`, after the regex capture we'd have `500`, which is an int, same as the whole number version we'd get.  I'll do what you're describing though.

Reply:
no here you converted `500m` to `0.5`, then ceil it and convert it to `1` (all in one line). what I said is to keep `0.5` instead.